### PR TITLE
feat: support gradient line for Progress

### PR DESCRIPTION
**Pull request by orzyyyy**
First of all, thank you for your contribution! 😄

New feature please send pull request to feature branch, and rest to master branch.
Pull request will be merged after one of collaborators approve.
Please makes sure that these form are filled before submitting your pull request, thank you!

[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Branch merge
- [ ] Other (about what?)

### 👻 What's the background?

1. Describe the source of requirement, like related issue link.

2. Describe the problem and the scenario.

### 💡 Solution

1. How to fix the problem, and list final API implementation and usage sample if that is an new feature.

2. GIF or snapshot should be provided if includes UI/interactive modification.
demo

    Feel free to close this if it's not in design.

### 📝 Changelog description

> Describe changes from userside, and list all potential break changes or other risks.

1. English description

    Support gradient line color for Progress.Line
2. Chinese description (optional)

    Progress.Line 支持渐变色
### ☑️ Self Check before Merge

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

**Comment by afc163** [comment]
这个好酷！

例子要找个好看点的配色

File: components/progress/demo/gradient-line.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+---
+order: 11
+title:
+  zh-CN: 自定义进度条渐变色
+  en-US: Custom line gradient
+---
+
+## zh-CN
+
+`linear-gradient` 的一封装。推荐只传两种颜色。
```
Comment (afc163) [inline components/progress/demo/gradient-line.md]: 一

Comment (zy410419243) [reply components/progress/demo/gradient-line.md]: ...好像没毛病啊，“这就一老实人。”

Comment (zy410419243) [reply components/progress/demo/gradient-line.md]: 哦太口语化了，我改改

File: components/progress/__tests__/__snapshots__/index.test.js.snap [hunk 2]
```diff
@@ -245,3 +245,47 @@ exports[`Progress render strokeColor 1`] = `
...
+exports[`Progress render strokeColor 2`] = `
+<div
+  class="ant-progress ant-progress-circle ant-progress-status-normal ant-progress-show-info ant-progress-default"
+>
+  <div
+    class="ant-progress-inner"
+    style="width: 120px; height: 120px; font-size: 24px;"
+  >
+    <svg
+      class="ant-progress-circle "
```
Comment (afc163) [inline components/progress/__tests__/__snapshots__/index.test.js.snap]: 怎么是 circle？

File: components/progress/Line.tsx [hunk 3]
```diff
@@ -18,11 +18,20 @@ const Line: React.SFC<LineProps> = props => {
     strokeLinecap,
     children,
   } = props;
+  let backgroundProps;
+  if (strokeColor && typeof strokeColor !== 'string') {
+    const { from = '#108ee9', to = '#87d068', direction = 'to right' } = strokeColor;
```
Comment (zombieJ) [inline components/progress/Line.tsx]: 不要写死 `from` / `to`，也要允许用户这么写 `'0%': 'red', '60%': 'green'`

Comment (zy410419243) [reply components/progress/Line.tsx]: 多段比较少见吧，一定要写也能写成
```jsx
const strokeColor = {
  from: 'red 0%, green 60%',
  to: 'black 100%',
}
```

Comment (zy410419243) [reply components/progress/Line.tsx]: 哦说的是写法，我改改...

File: components/progress/Line.tsx [hunk 4]
```diff
@@ -1,11 +1,27 @@
...
+import { ProgressProps, ProgressGradient } from './progress';
 
 interface LineProps extends ProgressProps {
   prefixCls: string;
   children: React.ReactNode;
 }
+type StringGradients = { [percent: string]: string };
+
+export const handleGradient = (strokeColor: ProgressGradient) => {
+  const { from = '#108ee9', to = '#87d068', direction = 'to right', ...rest } = strokeColor;
```
Comment (zombieJ) [inline components/progress/Line.tsx]: 默认颜色就用 primary 到 primary 吧，例子里写不一样的就行了。

File: components/progress/__tests__/index.test.js [hunk 5]
```diff
@@ -56,6 +56,13 @@ describe('Progress', () => {
       type: 'line',
     });
     expect(wrapper).toMatchSnapshot();
+    wrapper.setProps({
+      strokeColor: {
+        '0%': '#108ee9',
+        '100%': '#87d068',
+      },
```
Comment (zombieJ) [inline components/progress/__tests__/index.test.js]: 用户如果写成
```jsx
{
  '100%': 'xxx',
  '0%': 'xxx',
  '33%': 'xxxx',
}
```
形式，会不会有问题？

Comment (zy410419243) [reply components/progress/__tests__/index.test.js]: 这么写会原样输出，但万一人家就是要这样的怎么办...

Comment (zombieJ) [reply components/progress/__tests__/index.test.js]: 我们应该帮忙处理这种情况，又比如下面这种也是一个效果：
```jsx
const linerColor = { '100%': 'xxxx' };
if (xxx) {
  linerColor['0%'] = 'xxx';
} else {
  linerColor['50%'] = 'xxx';
}
```

Comment (zy410419243) [reply components/progress/__tests__/index.test.js]: 或者应该在 `strokeColor` 再加个原样输出的参数，然后默认排序？

Comment (zombieJ) [reply components/progress/__tests__/index.test.js]: 我们帮忙处理就行了

**Comment by afc163** [comment]
给几个配色参考：

https://codepen.io/laurenglegg/pen/ayEpVe
https://codepen.io/nikitahl/pen/XEzbNe
https://codepen.io/Travisaurus-Rex/pen/JRPLmm

另外能不能支持 circle？

**Comment by zy410419243** [comment]
文档里就红绿蓝三种颜色，颜色太多了看起来很累

我是觉着这颜色挺好的...

**Comment by afc163** [comment]
This pull request **introduces 1 alert** when merging 2d9cc6f01abff77907d97c42661550172fd40d04 into addc7bd26608e6057076708e6af00c37c732dbeb - [view on LGTM.com](https://lgtm.com/projects/g/ant-design/ant-design/rev/pr-27de17dbd5c47fb4e5924e9865e61bcbd9a3f95d)

**new alerts:**

* 1 for Incomplete string escaping or encoding

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: components/progress/Line.tsx [hunk 6]
```diff
@@ -1,11 +1,72 @@
...
+ *   '50%': 'green',     ====>     '50%': '#00CC00',
+ *   '25%': '#66FF00',             '75%': '#009900',
+ *   '100%': '#ffffff'             '100%': '#ffffff'
+ * }                             }
+ */
+export const sortGradient = (gradients: StringGradients) => {
+  let tempArr = [];
+  const result: StringGradients = {};
+  for (const [key, value] of Object.entries(gradients)) {
+    const formatKey = parseFloat(key.replace('%', ''));
```
Comment (afc163) [inline components/progress/Line.tsx]:
```suggestion
    const formatKey = parseFloat(key.replace(/%/g, ''));
```

https://lgtm.com/projects/g/ant-design/ant-design/rev/pr-27de17dbd5c47fb4e5924e9865e61bcbd9a3f95d

**Comment by dengfuping** [comment]
@zy410419243 Could you rebase the commits?

**Comment by zy410419243** [comment]
@dengfuping Feel free to squash merge, all is related commits.

File: components/progress/progress.tsx [hunk 7]
```diff
@@ -12,7 +12,11 @@ const ProgressTypes = tuple('line', 'circle', 'dashboard');
 export type ProgressType = (typeof ProgressTypes)[number];
 const ProgressStatuses = tuple('normal', 'exception', 'active', 'success');
 export type ProgressSize = 'default' | 'small';
-
+export type ProgressGradient = {
+  from: string;
+  to: string;
+  direction: string;
+};
```
Comment (zombieJ) [inline components/progress/progress.tsx]: 这里拆出 2 个，一个支持 form/to，一个支持 [percentage]

File: components/progress/__tests__/index.test.js [hunk 8]
```diff
@@ -48,10 +49,43 @@ describe('Progress', () => {
...
+      'linear-gradient(to right, test 0%)',
+    );
+  });
+
+  it('sort gradients correctly', () => {
+    expect(sortGradient({ '10%': 'test10', '30%': 'test30', '20%': 'test20' })).toEqual({
+      '10%': 'test10',
+      '20%': 'test20',
+      '30%': 'test30',
+    });
```
Comment (zombieJ) [inline components/progress/__tests__/index.test.js]: 这个测试随便顺序怎么样都是过的……

File: components/progress/Line.tsx [hunk 9]
```diff
@@ -1,11 +1,72 @@
...
+      key: formatKey,
+      value,
+    });
+  }
+  tempArr = tempArr.sort((a, b) => a.key - b.key);
+  for (const item of tempArr) {
+    const { key, value } = item;
+    result[key + '%'] = value;
+  }
+  return result;
```
Comment (zombieJ) [inline components/progress/Line.tsx]: 不要产出 object，sort 后直接产出 string。

File: components/progress/Line.tsx [hunk 10]
```diff
@@ -56,14 +55,8 @@ export const sortGradient = (gradients: StringGradients) => {
...
   if (Object.keys(rest).length !== 0) {
-    let backgroundImage = `linear-gradient(${direction}, `;
-    const sortedGradients = sortGradient(rest as StringGradients);
-    for (const key of Object.keys(sortedGradients)) {
-      const value = (rest as StringGradients)[key];
-      backgroundImage += `${value} ${key}, `;
-    }
-    backgroundImage = backgroundImage.substring(0, backgroundImage.length - 2) + ')';
-    return { backgroundImage };
+    const sortedGradients = sortGradient(rest as ProgressGradient);
```
Comment (zombieJ) [inline components/progress/Line.tsx]: 这边也可以拆出来了：StringGradients | FromToGradients

File: components/progress/Line.tsx [hunk 11]
```diff
@@ -33,9 +32,9 @@ export const sortGradient = (gradients: StringGradients) => {
   tempArr = tempArr.sort((a, b) => a.key - b.key);
   for (const item of tempArr) {
     const { key, value } = item;
-    result[key + '%'] = value;
+    result += `${value} ${key}%, `;
   }
-  return result;
+  return result.substring(0, result.length - 2);
```
Comment (zombieJ) [inline components/progress/Line.tsx]: ```tempArr.map(({ key, value }) => `${value} ${key}%`).join(', ');```?

Comment (zy410419243) [reply components/progress/Line.tsx]: 感谢，一行代码直接改了我一个坏习惯...

**Review by afc163 — APPROVED** [review]

**Pull request merged**